### ISSUE TITLE
Nullable reference types are not generated (doNotForceIsNullable not specified)

### ISSUE THREAD

**Issue by EshG** [mganss/XmlSchemaClassGenerator#270]
Are Nullable reference types constraints supported?

If it is, then by reading the help:

> do not force generator to emit IsNullable = true
>                                in XmlElement annotation for nillable elements
>                                when element is nullable (minOccurs < 1 or
>                                parent element is choice) (default is false)

I understand that without specifying the doNotForceIsNullable parameter, I should get a nullable attributes?

But generating from this xsd will not add any nullable declerations:

```
 <xs:complexType name="Child">
        <xs:sequence>
          <xs:element name="NullableA" type="CType" minOccurs="0" maxOccurs="1" />
          <xs:element name="NullableB" type="CType" minOccurs="0" maxOccurs="1" />
        </xs:sequence>
  </xs:complexType>


  <xs:complexType name="CType">
    </xs:complexType> 
```

I will be be glad to take a look if this is indeed a bug that should be fixed

**Comment by EshG** [comment]
CLI Version 2.0.560.0
Nullable ref is enabled at the project level.

**Comment by mganss** [comment]
Currently there is no support for nullable reference types. The `doNotForceIsNullable` option is about the [`IsNullable` attribute](https://docs.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlelementattribute.isnullable?view=net-5.0).

Would you like to take a crack at this feature?

**Comment by mganss** [comment]
> CLI Version 2.0.560.0
> Nullable ref is enabled at the project level.

🤔 No, not that I remember.

**Comment by EshG** [comment]
> Currently there is no support for nullable reference types. The `doNotForceIsNullable` option is about the [`IsNullable` attribute](https://docs.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlelementattribute.isnullable?view=net-5.0).
> 
> Would you like to take a crack at this feature?

Thanks for the clarification regarding dnfin. I'll be glad to get to that.  do you have any preferences / guidance for first time contributors?

**Comment by mganss** [comment]
@EshG Thanks. There are no specific rules, just try not to stray to far from the style used in the current codebase. A few things to watch out for when introducing a new command line parameter:

- In the unit tests, make sure to also add your new parameter here: https://github.com/mganss/XmlSchemaClassGenerator/blob/85ead03e87d96e134fc29234a5edbf4e730bebc3/XmlSchemaClassGenerator.Tests/Compiler.cs#L70-L126
- Due to the current design, you'll have to add the new configuration option in the classes `Generator`, `GeneratorConfiguration`, and `Program` (the actual CLI).
- When creating a unit test, just copy one of the existing tests that test generation, compilation, and de/serialization like e.g. the following and use your new configuration option: https://github.com/mganss/XmlSchemaClassGenerator/blob/85ead03e87d96e134fc29234a5edbf4e730bebc3/XmlSchemaClassGenerator.Tests/XmlTests.cs#L398-L404

**Comment by EshG** [comment]
@mganss Very well. I will try to send a basic PR when I have something to make sure I am aligned with you.

**Comment by EshG** [comment]
@mganss it appears that the project will need a .net upgrade to support this future. which version if any would you like the upgrade to?

**Comment by mganss** [comment]
@EshG This issue is only about the generated code right? The project's target framework shouldn't matter for this should it?

**Comment by EshG** [comment]
@mganss I agree, we can append text to the output. but doing so, in my opinion, will complicate the existing code  way too much. I also prefer to respect the way you do this using CodeDom which means adding the attributes as .net types.

**Comment by mganss** [comment]
@EshG Do you plan on adding attributes to properties to signal nullability? IIRC these can be added without having a reference to the attribute type by specifying the attribute's type as a string.

**Comment by EshG** [comment]
> Do you plan on adding attributes to properties to signal nullability?

Exactly.



> `IIRC these can be added without having a reference to the attribute type by specifying the attribute's type as a string.


Fair enough, I will take a look .

**Comment by EshG** [comment]
@mganss Apologies, but for now I won't be able to finish this this task.
Until next time, best of luck.

**Comment by AndreasHoffmann2** [comment]
I just had a look into the code if I can help with this.
Since this project uses System.CodeDom, there is no way to generate the proper syntax (e.g. "string?") as long as this project targets .NET Frameword and not .NET Standard. Is this correct?
You were writing about adding Attributes above. Is this a way to add the nullable-hints without using the syntax? How does it work?

**Comment by AndreasHoffmann2** [comment]
Well, I gave it a shot: 
https://github.com/froebel/XmlSchemaClassGenerator/commit/4bfc2c46ebfbe5839d5afebd855ffe2f70be5efa

I added an "AllowNullAttribute" to nullable reference types.

But since I don't know the project in depth, I don't know if I missed anything.

Further remarks:
- Unfortunately, this attribute does not seem to be available using reflection on the compiled assembly. Thus I had to parse the SyntaxTree. And to get the SyntaxTree, I needed to get it from Compiler.cs.
- I did not yet invest the effort to add the feature to the Console. Before that, I would like to get some feedback if this is the right direction.

**Comment by mganss** [comment]
@AndreasHoffmann2 Thanks a lot for your effort. Took me a while to figure it out, but I think I know what's going on with the `AllowNullAttribute`. ILSpy shows this:

```c#
public class AttributeReferenceNullable
{
	[XmlAttribute("text")]
	public string Text
	{
		[CompilerGenerated]
		get
		{
			return <Text>k__BackingField;
		}
		[CompilerGenerated]
		[param: AllowNull]
		set
		{
			<Text>k__BackingField = value;
		}
	}
}
```

You can get at the attribute through reflection like so:

```c#
var c = assembly.GetType("Test.ElementReferenceNullable");
var p = c.GetProperty("Text");
var s = p.SetMethod;
var ps = s.GetParameters();
var attr = ps[0].CustomAttributes.Single(a => a.AttributeType == typeof(System.Diagnostics.CodeAnalysis.AllowNullAttribute));
```

Could you refactor the code accordingly and open a PR?

Shouldn't the properties also get [`MaybeNullAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.maybenullattribute)?

Unfortunately, it seems that adding nullable attributes doesn't cause Intellisense to show the property's type as `string?` for example.

**Comment by AndreasHoffmann2** [comment]
Thank you for the research about the attribute. It did not occur to me to search for the attribute on the set-method. I'll change the code accordingly.
And I'll have a closer look into `MaybeNullAttribute`. In the first draft, I used this one but switched to `AllowNullAttribute` which seemed more appropriate. But since the attributes are added to the get-set-methods themselves it seems logical that one will be applied to the get and one to the set. I'll check this.

And for the moment, I'll leave the missing question-mark (`string?`) for later.

**Comment by mganss** [comment]
This has been fixed through #312
